### PR TITLE
Transform `processingElement` into a `Circuit`

### PR DESCRIPTION
**Pull request by lmbollen**
See #282

### CODE REVIEW

File: bittide/src/Bittide/ProcessingElement.hs [hunk 1]
```diff
@@ -40,42 +42,56 @@ processingElement ::
...
+   where
+    (fwdLeft, fwdRight) = splitAtI fwd
+    bwd = bwdLeft ++ bwdRight
+
+rvCircuit ::
+  (HiddenClockResetEnable dom) =>
+  Signal dom Bit ->
+  Signal dom Bit ->
+  Signal dom Bit ->
+  Circuit () (Wishbone dom 'Standard 32 (Bytes 4), Wishbone dom 'Standard 32 (Bytes 4))
```
Comment (cuddlefishie) [inline bittide/src/Bittide/ProcessingElement.hs:79]: More of a philosophical question than a review, but when should we use `Signal dom X -> Circuit () ...` vs `Circuit (CSignal dom X) ...`?

File: bittide/src/Bittide/DoubleBufferedRam.hs [hunk 2]
```diff
@@ -89,6 +89,32 @@ contentGenerator content = case compareSNat d1 (SNat @romSize) of
...
+  forall dom depth awA awB .
+  ( HiddenClockResetEnable dom
+  , KnownNat awA, 2 <= awA
+  , KnownNat awB, 2 <= awB
+  , KnownNat depth, 1 <= depth) =>
+  InitialContent depth (Bytes 4) ->
+  Circuit
+  ( Wishbone dom 'Standard awA (Bytes 4)
+  , Wishbone dom 'Standard awB (Bytes 4)
+  ) ()
```
Comment (cuddlefishie) [inline bittide/src/Bittide/DoubleBufferedRam.hs:102]:
```suggestion
  Circuit
    ( Wishbone dom 'Standard awA (Bytes 4)
    , Wishbone dom 'Standard awB (Bytes 4) )
    ()
```
or maybe (it's still below 90 chars wide)
```suggestion
  Circuit
    (Wishbone dom 'Standard awA (Bytes 4), Wishbone dom 'Standard awB (Bytes 4))
    ()
```

File: bittide/src/Bittide/ProcessingElement.hs (hunk 1 above)
Comment (lmbollen) [reply bittide/src/Bittide/ProcessingElement.hs:79]: That's a great question and the answer is a strong: _I don´t know yet_. Perhaps @martijnbastiaan does.
I can imagine we´d want to rewrite this to:
```hs
rvCircuit ::
  (HiddenClockResetEnable dom) =>
  Circuit 
    (CSignal dom Bit, CSignal dom Bit, CSignal dom Bit) 
    (Wishbone dom 'Standard 32 (Bytes 4), Wishbone dom 'Standard 32 (Bytes 4))
```

File: bittide/src/Bittide/ProcessingElement.hs [hunk 3]
```diff
@@ -2,17 +2,19 @@
 --
 -- SPDX-License-Identifier: Apache-2.0
 
+{-# LANGUAGE BlockArguments #-}
```
[the remark this answers is not in the document]
Comment (lmbollen) [reply bittide/src/Bittide/ProcessingElement.hs:5]: Now it does ;)

File: bittide/src/Bittide/ProcessingElement.hs (hunk 1 above)
Comment (martijnbastiaan) [reply bittide/src/Bittide/ProcessingElement.hs:80]: Using `CSignal` outside of a `Circuit` means that you can't use `|>` to compose anymore, provided that you have another `Circuit` with the right shape. Other than that there is no difference AFAIK. (Perhaps the plugin doesn't understand if you put it outside?) I'd slightly prefer yeeting everything in `Circuit`.

I also think we should use vertical space a bit more aggressively; just treat them as arguments / results, so:

```haskell
  Circuit
    ( CSignal dom Bit
    , CSignal dom Bit
    , CSignal dom Bit
    ) 
    ( Wishbone dom 'Standard 32 (Bytes 4)
    , Wishbone dom 'Standard 32 (Bytes 4)
    )
```

makes it much more obvious how many buses the circuit has. (Alternatively, the trailing `)` can go on the previous line. I don't mind.)

File: bittide/src/Bittide/DoubleBufferedRam.hs [hunk 4]
```diff
@@ -493,10 +518,10 @@ splitWriteInBytes ::
   Vec (Regs writeData 8) (Maybe (LocatedByte maxIndex))
 splitWriteInBytes (Just (addr, writeData)) byteSelect =
   case getRegsBe writeData of
-    RegisterBank vec -> splitWrites <$> unpack byteSelect <*> vec
+    RegisterBank vec -> orNothing <$> unpack byteSelect <*> fmap (addr,) vec
      where
-      splitWrites :: Bool -> Byte -> Maybe (LocatedByte maxIndex)
-      splitWrites b bv = orNothing b (addr, bv)
+      -- splitWrites :: Bool -> Byte -> Maybe (LocatedByte maxIndex)
+      -- splitWrites b bv = if b then Just (addr, bv) else Nothing
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/DoubleBufferedRam.hs:524]: Remove?

File: bittide/src/Bittide/Node.hs [hunk 5]
```diff
@@ -151,7 +152,7 @@ gppe (GppeConfig linkConfig peConfig, linkIn, splitAtI -> (nmuM2S0, nmuM2S1)) =
  where
   (suS2M, nmuS2M0) = linkToPe linkConfig linkIn sc suM2S nmuM2S0
   (linkOut, guS2M, nmuS2M1) = peToLink linkConfig sc guM2S nmuM2S1
-  (suM2S :> guM2S :> Nil) = processingElement peConfig (suS2M :> guS2M :> Nil)
+  (_,suM2S :> guM2S :> Nil) = toSignals (processingElement peConfig) ((),suS2M :> guS2M :> Nil)
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/Node.hs:155]:
```suggestion
  (_, suM2S :> guM2S :> Nil) = toSignals (processingElement peConfig) ((),suS2M :> guS2M :> Nil)
```

File: bittide/src/Bittide/Node.hs [hunk 6]
```diff
@@ -182,6 +183,6 @@ managementUnit (ManagementConfig linkConfig peConfig) linkIn nodeS2Ms =
   (linkOut, guS2M, nmuS2M1) = peToLink linkConfig sc guM2S nmuM2S1
   (suM2S :> guM2S :> rest) = nmuM2Ss
   (splitAtI -> (nmuM2S0, nmuM2S1), nodeM2Ss) = splitAtI rest
-  nmuM2Ss = processingElement peConfig nmuS2Ms
+  (_,nmuM2Ss) = toSignals (processingElement peConfig) ((),nmuS2Ms)
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/Node.hs:186]:
```suggestion
  (_, nmuM2Ss) = toSignals (processingElement peConfig) ((),nmuS2Ms)
```

Won't someone think of the consistency :cry:

File: bittide/src/Bittide/ProcessingElement.hs [hunk 7]
```diff
@@ -41,42 +43,56 @@ processingElement ::
...
   , KnownNat nBusses, 2 <= nBusses, CLog 2 nBusses <= 30) =>
   PeConfig nBusses ->
-  Vec (nBusses-2) (Signal dom (WishboneS2M (Bytes 4))) ->
-  Vec (nBusses-2) (Signal dom (WishboneM2S (32 - CLog 2 nBusses) 4 (Bytes 4)))
-processingElement config bussesIn = case config of
-  PeConfig memMapConfig initI initD ->
-    go memMapConfig initI initD
+  Circuit () (Vec (nBusses-2) (Wishbone dom 'Standard (MappedBus 32 nBusses) (Bytes 4)))
+processingElement (PeConfig memMapConfig' initI' initD') =
+  go memMapConfig' initI' initD'
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ProcessingElement.hs:48]: Why is there a  `go`? Seems like the only thing you're doing is unwrapping the record?

File: bittide/src/Bittide/DoubleBufferedRam.hs [hunk 8]
```diff
@@ -89,6 +89,31 @@ contentGenerator content = case compareSNat d1 (SNat @romSize) of
   _ -> (pure Nothing, pure True)
 
 
+wbStorageDPC ::
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/DoubleBufferedRam.hs:92]: Haddock pls

```suggestion
wbStorageDPC ::
```

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]

File: bittide/src/Bittide/ProcessingElement.hs (hunk 7 above)
Comment (lmbollen) [reply bittide/src/Bittide/ProcessingElement.hs:48]: `PeConfig nBusses` is a GADT, not a record :). This structure enables us to hide all type variables that are not relevant outside of `processingElement` from the API of `processingElement`.

File: bittide/src/Bittide/ProcessingElement.hs (hunk 1 above)
Comment (lmbollen) [reply bittide/src/Bittide/ProcessingElement.hs:80]: I feel that does make it more difficult to partially apply cirucits.

e.g.
With the current implementation you can simply write:
```hs
rv' = rvCircuit (pure 0) (pure 0) (pure 0)
```
to partially apply `rvCircuit`.

Comment (martijnbastiaan) [reply bittide/src/Bittide/ProcessingElement.hs]: Makes sense

Comment (martijnbastiaan) [reply bittide/src/Bittide/ProcessingElement.hs]: Any idea what the circuit plugin does?

File: bittide/src/Bittide/ProcessingElement.hs (hunk 7 above)
Comment (martijnbastiaan) [reply bittide/src/Bittide/ProcessingElement.hs:47]: Hmmm, let's discuss this in the office :-). Not sure what you're getting at.

**Comment by martijnbastiaan** [comment]
There are a few review suggestions you haven't touched yet, I'm going to remove myself as a reviewer to get it off my TODO list. Feel free to retag.

File: bittide/src/Bittide/ProcessingElement.hs (hunk 7 above)
Comment (lmbollen) [reply bittide/src/Bittide/ProcessingElement.hs]: It turns out that the go function was not necessary to hide the type variables, pattern matching on the GADT is enough.

File: bittide/src/Bittide/ProcessingElement.hs [hunk 9]
```diff
@@ -41,44 +42,53 @@ processingElement ::
...
+  Circuit () (Vec (nBusses-2) (Wishbone dom 'Standard (MappedBus 32 nBusses) (Bytes 4)))
+processingElement (PeConfig memMapConfig initI initD) = circuit $ do
+  (iBus, dBus) <- rvCircuit (pure low) (pure low) (pure low)
+  ([iMemBus, dMemBus], extBusses) <-
+    (splitAtCircuit d2 <| singleMasterInterconnect memMapConfig) -< dBus
+  wbStorage initI -< iBus
+  wbStorageDPC initD -< (iMemBus, dMemBus)
+  idC -< extBusses
+
+splitAtCircuit ::
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ProcessingElement.hs:54]:
```suggestion
-- | Conceptually the same as 'splitAt', but for 'Circuit's
splitAtC ::
```

Comment (martijnbastiaan) [reply bittide/src/Bittide/ProcessingElement.hs:54]: I think we suffix all the other circuit specific helpers with `C` in `clash-protocols`.

**Review by martijnbastiaan — APPROVED** [review]
LGTM. Don't merge before looking at the last comment please :).

**Pull request merged**